test/fuzz/CMakeLists.txt

```diff
@@ -1,30 +1,109 @@
 #
-# Copyright (c) 2023 alandefreitas (alandefreitas@gmail.com)
+# Copyright (c) 2023 Alan de Freitas (alandefreitas@gmail.com)
 #
 # Distributed under the Boost Software License, Version 1.0.
 # https://www.boost.org/LICENSE_1_0.txt
 #
 
-# Files
-source_group("" FILES fuzz_parse.cpp)
-
-function(add_boost_url_fuzzer NAME SOURCE_FILES)
-    # Fuzzer library
-    add_library(fuzzerlib_${NAME} ${SOURCE_FILES})
-    target_link_libraries(fuzzerlib_${NAME} PRIVATE Boost::url)
-    set_property(TARGET fuzzerlib_${NAME} PROPERTY FOLDER "fuzzing")
-
-    if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-        # Fuzzer executable
-        add_executable(fuzzer_${NAME} ${SOURCE_FILES})
-        target_link_libraries(fuzzer_${NAME} PRIVATE Boost::url)
-        target_compile_options(fuzzer_${NAME} PRIVATE -g -O2 -fsanitize=fuzzer,address,undefined -fno-sanitize-recover=undefined)
-        target_link_libraries(fuzzer_${NAME} PRIVATE -fsanitize=fuzzer -fuse-ld=lld)
-
-        # Custom target to run fuzzer executable
-        add_custom_target(fuzz_${NAME} fuzzer_${NAME} -rss_limit_mb=8192 -max_total_time=30 -timeout=30 DEPENDS fuzz_${NAME})
-        set_property(TARGET fuzzer_${NAME} PROPERTY FOLDER "fuzzing")
-    endif ()
+# Get number of cores
+include(ProcessorCount)
+ProcessorCount(PROCESSOR_COUNT)
+
+# Determine total fuzz time per file
+file(GLOB BOOST_URL_FUZZER_SOURCE_FILES *.cpp)
+list(LENGTH BOOST_URL_FUZZER_SOURCE_FILES BOOST_URL_FUZZER_SOURCE_FILES_COUNT)
+set(BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT 30)
+math(EXPR BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT "${BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT} / (${PROCESSOR_COUNT} * ${BOOST_URL_FUZZER_SOURCE_FILES_COUNT}) + 1")
+
+# Fuzzing options
+set(BOOST_URL_FUZZER_TOTAL_TIME ${BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT} CACHE STRING "Total time for fuzzing")
+set(BOOST_URL_FUZZER_RSS_LIMIT 8192 CACHE STRING "RSS limit for fuzzing")
+set(BOOST_URL_FUZZER_TIMEOUT 30 CACHE STRING "Timeout for fuzzing")
+set(BOOST_URL_FUZZER_MAX_LEN 4000 CACHE STRING "Maximum size of the input")
+set(BOOST_URL_FUZZER_JOBS ${PROCESSOR_COUNT} CACHE STRING "Number of jobs for fuzzing")
+option(BOOST_URL_FUZZER_ADD_TO_CTEST "Add fuzzing targets to ctest" OFF)
+
+# Corpus
+set(BOOST_URL_FUZZER_CORPUS_PATH ${CMAKE_CURRENT_SOURCE_DIR}/corpus.tar)
+set(BOOST_URL_FUZZER_CORPUS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/corpus)
+if(EXISTS BOOST_URL_FUZZER_CORPUS_PATH)
+    add_custom_target(
+        untar_corpus
+        COMMAND ${CMAKE_COMMAND} -E echo "Untar corpus from ${BOOST_URL_FUZZER_CORPUS_PATH}"
+        COMMAND ${CMAKE_COMMAND} -E tar xvf ${BOOST_URL_FUZZER_CORPUS_PATH}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+        COMMENT "Unzipping corpus"
+        VERBATIM)
+else()
+    add_custom_target(untar_corpus
+        COMMAND ${CMAKE_COMMAND} -E echo "No corpus. Create ${CMAKE_CURRENT_SOURCE_DIR}/corpus."
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_CURRENT_SOURCE_DIR}/corpus
```

Comment:
Yes. I got this pattern from the b2 scripts and thought this would help me keep them compatible. But it doesn't make sense now that I removed it from b2. 

Moving the corpus to another dir for the corpus is easy because I'm already in CMake and can use CMAKE_CURRENT_SOURCE_DIR. But I'm sure where to put the initial corpus.tar from the cache before running cmake. What do you recommend?

Reply:
Do the initial seeds in `test/fuzz/seeds` look OK?

Reply:
Yes they do. I store them as a single, compressed file bc I've got the feeling it's faster to checkout (mine are binary files tho).

The best way to check whether they're okay is measuring coverage though. You can follow [this](https://docs.cppalliance.org/contributor-guide/testing/fuzzing.html#_verifying_the_effectiveness_of_your_fuzzer) to do it.

Reply:
Oh! So site-docs is already being useful. :)

Reply:
OK. I found a solution to this. The corpus goes to `${CMAKE_CURRENT_BINARY_DIR}/corpus` and `corpus.tar` comes from an option `BOOST_URL_FUZZER_CORPUS_PATH` defaulting to `${CMAKE_CURRENT_BINARY_DIR}/corpus.tar`. 
This means it can locally keep and reuse a corpus.tar in the build dir and in CI it downloads the cached corpus somewhere and sets the option.

I also put the seeds in a single archive that's extracted to `${CMAKE_CURRENT_BINARY_DIR}/seeds` when running the target.